File: app/modules/settings/schemas.py
from __future__ import annotations

from pydantic import Field, field_validator

from app.modules.shared.schemas import DashboardModel

_DEFAULT_WEEKLY_PACE_WORKING_DAYS = "0,1,2,3,4,5,6"
_WEEKLY_PACE_SMOOTHING_MINUTES = (15, 30, 60, 120, 240)
_HTTP_DOWNSTREAM_TRANSPORT_POLICY_PATTERN = r"^(smart|always_http|always_websocket|pinned)$"


def _normalize_weekly_pace_working_days(value: str | None) -> str | None:
    if value is None:
        return None
    tokens = [part.strip() for part in value.split(",") if part.strip()]
    if not tokens:
        raise ValueError("weekly_pace_working_days must include at least one day")
    try:
        days = sorted({int(token) for token in tokens})
    except ValueError as exc:
        raise ValueError("weekly_pace_working_days must contain weekday numbers") from exc
    if any(day < 0 or day > 6 for day in days):
        raise ValueError("weekly_pace_working_days must use 0-6 weekday numbers")
    return ",".join(str(day) for day in days)


class AdditionalQuotaPolicy(DashboardModel):
    quota_key: str
    display_label: str
    routing_policy: str = Field(pattern=r"^(inherit|burn_first|normal|preserve)$")
    model_ids: list[str] = Field(default_factory=list)


class DashboardSettingsResponse(DashboardModel):
    sticky_threads_enabled: bool
    upstream_stream_transport: str = Field(pattern=r"^(default|auto|http|websocket)$")
    prohibit_fast_mode: bool
    http_downstream_transport_policy: str = Field(pattern=_HTTP_DOWNSTREAM_TRANSPORT_POLICY_PATTERN)
    proxy_account_response_create_limit: int = Field(ge=0)
    proxy_account_stream_limit: int = Field(ge=0)
    proxy_account_stream_recovery_reserve: int = Field(ge=0)
    proxy_api_key_fair_share_congestion_threshold_pct: int = Field(ge=0, le=100)
    upstream_proxy_routing_enabled: bool
    upstream_proxy_default_pool_id: str | None = None
    prefer_earlier_reset_accounts: bool
    prefer_earlier_reset_window: str = Field(pattern=r"^(primary|secondary)$")
    show_reset_credit_badges: bool
    auto_redeem_reset_credits_before_expiry: bool
    show_reset_credit_expiry_badge: bool
    routing_strategy: str = Field(
        pattern=r"^(usage_weighted|round_robin|capacity_weighted|relative_availability|fill_first|sequential_drain|reset_drain|single_account)$"
    )
    relative_availability_power: float = Field(gt=0.0)
    relative_availability_top_k: int = Field(ge=1, le=20)
    single_account_id: str | None = None
    openai_cache_affinity_max_age_seconds: int = Field(gt=0)
    dashboard_session_ttl_seconds: int = Field(ge=3600)
    http_responses_session_bridge_prompt_cache_idle_ttl_seconds: int = Field(gt=0)
    http_responses_session_bridge_gateway_safe_mode: bool
    sticky_reallocation_budget_threshold_pct: float = Field(ge=0.0, le=100.0)
    sticky_reallocation_primary_budget_threshold_pct: float = Field(ge=0.0, le=100.0)
    sticky_reallocation_secondary_budget_threshold_pct: float = Field(ge=0.0, le=100.0)
    warmup_model: str = Field(min_length=1)
    import_without_overwrite: bool
    totp_required_on_login: bool
    totp_configured: bool
    api_key_auth_enabled: bool
    hide_upstream_quota_from_api_keys: bool
    limit_warmup_enabled: bool
    limit_warmup_windows: str = Field(pattern=r"^(primary|secondary|both)$")
    limit_warmup_model: str = Field(min_length=1, max_length=128)
    limit_warmup_prompt: str = Field(min_length=1, max_length=512)
    limit_warmup_cooldown_seconds: int = Field(ge=60)
    limit_warmup_exhausted_threshold_percent: float = Field(gt=0.0, le=100.0)
    limit_warmup_idle_threshold_percent: float = Field(gt=0.0, le=100.0)
    limit_warmup_min_available_percent: float = Field(gt=0.0, le=100.0)
    weekly_pace_working_days: str = _DEFAULT_WEEKLY_PACE_WORKING_DAYS
    weekly_pace_smoothing_minutes: int = Field(default=30)
    limit_warmup_staggered_idle_enabled: bool
    request_log_retention_days: int = Field(ge=0, le=3650)
    usage_history_retention_days: int = Field(ge=0, le=3650)
    request_log_retention_override_days: int | None = Field(default=None, ge=0, le=3650)
    usage_history_retention_override_days: int | None = Field(default=None, ge=0, le=3650)
    additional_quota_routing_policies: dict[str, str] = Field(default_factory=dict)
    additional_quota_policies: list[AdditionalQuotaPolicy] = Field(default_factory=list)
    guest_access_enabled: bool
    guest_password_configured: bool
    version: int = Field(ge=1)


class DashboardSettingsUpdateRequest(DashboardModel):
    expected_version: int | None = Field(default=None, ge=1)
    sticky_threads_enabled: bool | None = None
    upstream_stream_transport: str | None = Field(
        default=None,
        pattern=r"^(default|auto|http|websocket)$",
    )
    prohibit_fast_mode: bool | None = None
    http_downstream_transport_policy: str | None = Field(
        default=None,
        pattern=_HTTP_DOWNSTREAM_TRANSPORT_POLICY_PATTERN,
    )
    proxy_account_response_create_limit: int | None = Field(default=None, ge=0)
    proxy_account_stream_limit: int | None = Field(default=None, ge=0)
    proxy_account_stream_recovery_reserve: int | None = Field(default=None, ge=0)
    proxy_api_key_fair_share_congestion_threshold_pct: int | None = Field(default=None, ge=0, le=100)
    upstream_proxy_routing_enabled: bool | None = None
    upstream_proxy_default_pool_id: str | None = None
    prefer_earlier_reset_accounts: bool | None = None
    prefer_earlier_reset_window: str | None = Field(default=None, pattern=r"^(primary|secondary)$")
    show_reset_credit_badges: bool | None = None
    auto_redeem_reset_credits_before_expiry: bool | None = None
    show_reset_credit_expiry_badge: bool | None = None
    routing_strategy: str | None = Field(
        default=None,
        pattern=r"^(usage_weighted|round_robin|capacity_weighted|relative_availability|fill_first|sequential_drain|reset_drain|single_account)$",
    )
    relative_availability_power: float | None = Field(default=None, gt=0.0)
    relative_availability_top_k: int | None = Field(default=None, ge=1, le=20)
    single_account_id: str | None = Field(default=None, max_length=255)
    openai_cache_affinity_max_age_seconds: int | None = Field(default=None, gt=0)
    dashboard_session_ttl_seconds: int | None = Field(default=None, ge=3600)
    http_responses_session_bridge_prompt_cache_idle_ttl_seconds: int | None = Field(default=None, gt=0)
    http_responses_session_bridge_gateway_safe_mode: bool | None = None
    sticky_reallocation_budget_threshold_pct: float | None = Field(default=None, ge=0.0, le=100.0)
    sticky_reallocation_primary_budget_threshold_pct: float | None = Field(default=None, ge=0.0, le=100.0)
    sticky_reallocation_secondary_budget_threshold_pct: float | None = Field(default=None, ge=0.0, le=100.0)
    additional_quota_routing_policies: dict[str, str] | None = None
    warmup_model: str | None = Field(default=None, min_length=1)
    import_without_overwrite: bool | None = None
    totp_required_on_login: bool | None = None
    api_key_auth_enabled: bool | None = None
    hide_upstream_quota_from_api_keys: bool | None = None
    limit_warmup_enabled: bool | None = None
    limit_warmup_windows: str | None = Field(default=None, pattern=r"^(primary|secondary|both)$")
    limit_warmup_model: str | None = Field(default=None, min_length=1, max_length=128)
    limit_warmup_prompt: str | None = Field(default=None, min_length=1, max_length=512)
    limit_warmup_cooldown_seconds: int | None = Field(default=None, ge=60)
    limit_warmup_exhausted_threshold_percent: float | None = Field(default=None, gt=0.0, le=100.0)
    limit_warmup_idle_threshold_percent: float | None = Field(default=None, gt=0.0, le=100.0)
    limit_warmup_min_available_percent: float | None = Field(default=None, gt=0.0, le=100.0)
    weekly_pace_working_days: str | None = None
    weekly_pace_smoothing_minutes: int | None = None
    guest_access_enabled: bool | None = None
    limit_warmup_staggered_idle_enabled: bool | None = None
    # Tri-state retention overrides: absent = unchanged, present null = clear
    # the override (back to inheriting the deprecated env alias), present
    # value = store the override.
    request_log_retention_override_days: int | None = Field(default=None, ge=0, le=3650)
    usage_history_retention_override_days: int | None = Field(default=None, ge=0, le=3650)

    @field_validator("request_log_retention_override_days")
    @classmethod
    def _validate_request_log_retention_override(cls, value: int | None) -> int | None:
        if value is not None and value != 0 and value < 30:
            raise ValueError("request_log_retention_override_days must be 0 (disabled) or >= 30")
        return value

    @field_validator("usage_history_retention_override_days")
    @classmethod
    def _validate_usage_history_retention_override(cls, value: int | None) -> int | None:
        if value is not None and value != 0 and value < 45:
            raise ValueError("usage_history_retention_override_days must be 0 (disabled) or >= 45")
        return value

    @field_validator("warmup_model")
    @classmethod
    def _normalize_warmup_model(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip()
        if not normalized:
            raise ValueError("warmup_model must not be blank")
        return normalized

    @field_validator("weekly_pace_working_days")
    @classmethod
    def _normalize_weekly_pace_days(cls, value: str | None) -> str | None:
        return _normalize_weekly_pace_working_days(value)

    @field_validator("weekly_pace_smoothing_minutes")
    @classmethod
    def _validate_weekly_pace_smoothing_minutes(cls, value: int | None) -> int | None:
        if value is None:
            return None
        if value not in _WEEKLY_PACE_SMOOTHING_MINUTES:
            raise ValueError("weekly_pace_smoothing_minutes must be one of 15, 30, 60, 120, 240")
        return value


class RuntimeConnectAddressResponse(DashboardModel):
    connect_address: str


class UpstreamProxyEndpointCreateRequest(DashboardModel):
    name: str = Field(min_length=1, max_length=128)
    scheme: str = Field(pattern=r"^(http|https|socks5|socks5h)$")
    host: str = Field(min_length=1, max_length=255)
    port: int = Field(ge=1, le=65535)
    username: str | None = Field(default=None, max_length=255)
    password: str | None = Field(default=None, max_length=1024)
    is_active: bool = True


class UpstreamProxyEndpointResponse(DashboardModel):
    id: str
    name: str
    scheme: str
    host: str
    port: int
    username: str | None
    is_active: bool


class UpstreamProxyEndpointTestResponse(DashboardModel):
    endpoint_id: str
    ok: bool
    status_code: int | None = None
    elapsed_ms: int | None = None
    error: str | None = None


class UpstreamProxyEndpointUpdateRequest(DashboardModel):
    name: str = Field(min_length=1, max_length=128)
    scheme: str = Field(pattern=r"^(http|https|socks5|socks5h)$")
    host: str = Field(min_length=1, max_length=255)
    port: int = Field(ge=1, le=65535)
    username: str | None = Field(default=None, max_length=255)
    # Omitted/null keeps the stored secret; a non-empty value replaces it.
    password: str | None = Field(default=None, max_length=1024)
    is_active: bool = True


class UpstreamProxyPoolCreateRequest(DashboardModel):
    name: str = Field(min_length=1, max_length=128)
    endpoint_ids: list[str] = Field(default_factory=list)
    is_active: bool = True


class UpstreamProxyPoolMemberRequest(DashboardModel):
    endpoint_id: str = Field(min_length=1)
    sort_order: int = 0
    weight: int = Field(default=1, ge=1)
    is_active: bool = True


class UpstreamProxyPoolUpdateRequest(DashboardModel):
    name: str = Field(min_length=1, max_length=128)
    endpoint_ids: list[str] = Field(default_factory=list)
    is_active: bool = True


class UpstreamProxyPoolResponse(DashboardModel):
    id: str
    name: str
    is_active: bool
    endpoint_ids: list[str]


class AccountProxyBindingRequest(DashboardModel):
    pool_id: str = Field(min_length=1)
    is_active: bool = True


class AccountProxyBindingResponse(DashboardModel):
    account_id: str
    pool_id: str
    is_active: bool


class UpstreamProxyAdminResponse(DashboardModel):
    routing_enabled: bool
    default_pool_id: str | None
    endpoints: list[UpstreamProxyEndpointResponse]
    pools: list[UpstreamProxyPoolResponse]
    bindings: list[AccountProxyBindingResponse]

File: frontend/src/features/settings/schemas.ts
import { z } from "zod";

const RoutingStrategySchema = z.enum([
  "usage_weighted",
  "round_robin",
  "capacity_weighted",
  "sequential_drain",
  "reset_drain",
  "single_account",
  "relative_availability",
  "fill_first",
]);
const UpstreamStreamTransportSchema = z.enum([
  "default",
  "auto",
  "http",
  "websocket",
]);
const HttpDownstreamTransportPolicySchema = z.enum([
  "smart",
  "always_http",
  "always_websocket",
  "pinned",
]);
const LimitWarmupWindowsSchema = z.enum([
  "primary",
  "secondary",
  "both",
]);
const AdditionalQuotaRoutingPolicySchema = z.enum([
  "inherit",
  "normal",
  "burn_first",
  "preserve",
]);
const AdditionalQuotaPolicySchema = z.object({
  quotaKey: z.string(),
  displayLabel: z.string(),
  routingPolicy: AdditionalQuotaRoutingPolicySchema,
  modelIds: z.array(z.string()).optional().default([]),
});
const LimitWarmupModelSchema = z.string().min(1).max(128);
const LimitWarmupPromptSchema = z.string().min(1).max(512);
const WeeklyPaceWorkingDaysValueSchema = z.string().regex(/^[0-6](,[0-6])*$/);
const WeeklyPaceWorkingDaysSchema = WeeklyPaceWorkingDaysValueSchema.default("0,1,2,3,4,5,6");
const WeeklyPaceSmoothingMinutesSchema = z.union([
  z.literal(15),
  z.literal(30),
  z.literal(60),
  z.literal(120),
  z.literal(240),
]);

export const DashboardSettingsSchema = z
  .object({
    stickyThreadsEnabled: z.boolean(),
    upstreamStreamTransport:
      UpstreamStreamTransportSchema.optional().default("default"),
    prohibitFastMode: z.boolean().optional().default(false),
    httpDownstreamTransportPolicy:
      HttpDownstreamTransportPolicySchema.optional().default("smart"),
    upstreamProxyRoutingEnabled: z.boolean().optional().default(false),
    upstreamProxyDefaultPoolId: z.string().nullable().optional().default(null),
    preferEarlierResetAccounts: z.boolean(),
    preferEarlierResetWindow: z.enum(["primary", "secondary"]).optional().default("secondary"),
    showResetCreditBadges: z.boolean().optional().default(true),
    autoRedeemResetCreditsBeforeExpiry: z.boolean().optional().default(false),
    showResetCreditExpiryBadge: z.boolean().optional().default(true),
    routingStrategy: RoutingStrategySchema.optional().default("usage_weighted"),
    relativeAvailabilityPower: z.number().positive().optional().default(2),
    relativeAvailabilityTopK: z
      .number()
      .int()
      .min(1)
      .max(20)
      .optional()
      .default(5),
    singleAccountId: z.string().nullable().optional().default(null),
    proxyAccountResponseCreateLimit: z.number().int().min(0).optional().default(4),
    proxyAccountStreamLimit: z.number().int().min(0).optional().default(8),
    proxyAccountStreamRecoveryReserve: z.number().int().min(0).optional().default(1),
    proxyApiKeyFairShareCongestionThresholdPct: z.number().int().min(0).max(100).optional().default(0),
    openaiCacheAffinityMaxAgeSeconds: z
      .number()
      .int()
      .positive()
      .optional()
      .default(300),
    dashboardSessionTtlSeconds: z
      .number()
      .int()
      .min(3600)
      .optional()
      .default(31536000),
    stickyReallocationBudgetThresholdPct: z.number().min(0).max(100).optional(),
    stickyReallocationPrimaryBudgetThresholdPct: z.number().min(0).max(100).optional(),
    stickyReallocationSecondaryBudgetThresholdPct: z.number().min(0).max(100).optional(),
    additionalQuotaRoutingPolicies: z
      .record(z.string(), AdditionalQuotaRoutingPolicySchema)
      .optional(),
    additionalQuotaPolicies: z.array(AdditionalQuotaPolicySchema).optional().default([]),
    warmupModel: z.string().trim().min(1).optional().default("gpt-5.4-mini"),
    importWithoutOverwrite: z.boolean(),
    totpRequiredOnLogin: z.boolean(),
    totpConfigured: z.boolean(),
    apiKeyAuthEnabled: z.boolean(),
    hideUpstreamQuotaFromApiKeys: z.boolean().optional().default(false),
    limitWarmupEnabled: z.boolean().optional().default(false),
    limitWarmupWindows: LimitWarmupWindowsSchema.optional().default("both"),
    limitWarmupModel: LimitWarmupModelSchema.optional().default("auto"),
    limitWarmupPrompt: LimitWarmupPromptSchema.optional().default("Say OK."),
    limitWarmupCooldownSeconds: z.number().int().min(60).optional().default(3600),
    limitWarmupExhaustedThresholdPercent: z
      .number()
      .positive()
      .max(100)
      .optional()
      .default(99),
    limitWarmupIdleThresholdPercent: z
      .number()
      .positive()
      .max(100)
      .optional()
      .default(1),
    limitWarmupMinAvailablePercent: z
      .number()
      .positive()
      .max(100)
      .optional()
      .default(100),
    weeklyPaceWorkingDays: WeeklyPaceWorkingDaysSchema,
    weeklyPaceSmoothingMinutes: WeeklyPaceSmoothingMinutesSchema.optional().default(30),
    guestAccessEnabled: z.boolean().optional().default(false),
    guestPasswordConfigured: z.boolean().optional().default(false),
    limitWarmupStaggeredIdleEnabled: z.boolean().optional().default(false),
    requestLogRetentionDays: z.number().int().min(0).max(3650).optional().default(0),
    usageHistoryRetentionDays: z.number().int().min(0).max(3650).optional().default(0),
    requestLogRetentionOverrideDays: z.number().int().min(0).max(3650).nullable().optional().default(null),
    usageHistoryRetentionOverrideDays: z.number().int().min(0).max(3650).nullable().optional().default(null),
    version: z.number().int().min(1).optional(),
  })
  .transform((settings) => {
    const legacyProvided = settings.stickyReallocationBudgetThresholdPct !== undefined;
    const primaryProvided = settings.stickyReallocationPrimaryBudgetThresholdPct !== undefined;
    const secondaryProvided = settings.stickyReallocationSecondaryBudgetThresholdPct !== undefined;
    const primaryThreshold =
      settings.stickyReallocationPrimaryBudgetThresholdPct ??
      settings.stickyReallocationBudgetThresholdPct ??
      95;
    return {
      ...settings,
      stickyReallocationBudgetThresholdPct:
        settings.stickyReallocationBudgetThresholdPct ?? primaryThreshold,
      stickyReallocationPrimaryBudgetThresholdPct: primaryThreshold,
      stickyReallocationSecondaryBudgetThresholdPct:
        settings.stickyReallocationSecondaryBudgetThresholdPct ??
        settings.stickyReallocationBudgetThresholdPct ??
        100,
      __stickyReallocationBudgetThresholdPctProvided: legacyProvided,
      __stickyReallocationPrimaryBudgetThresholdPctProvided: primaryProvided,
      __stickyReallocationSecondaryBudgetThresholdPctProvided: secondaryProvided,
    };
  });

export const SettingsUpdateRequestSchema = z
  .object({
    expectedVersion: z.number().int().min(1).optional(),
    stickyThreadsEnabled: z.boolean().optional(),
    upstreamStreamTransport: UpstreamStreamTransportSchema.optional(),
    prohibitFastMode: z.boolean().optional(),
    httpDownstreamTransportPolicy: HttpDownstreamTransportPolicySchema.optional(),
    upstreamProxyRoutingEnabled: z.boolean().optional(),
    upstreamProxyDefaultPoolId: z.string().nullable().optional(),
    preferEarlierResetAccounts: z.boolean().optional(),
    preferEarlierResetWindow: z.enum(["primary", "secondary"]).optional(),
    showResetCreditBadges: z.boolean().optional(),
    autoRedeemResetCreditsBeforeExpiry: z.boolean().optional(),
    showResetCreditExpiryBadge: z.boolean().optional(),
    routingStrategy: RoutingStrategySchema.optional(),
    relativeAvailabilityPower: z.number().positive().optional(),
    relativeAvailabilityTopK: z.number().int().min(1).max(20).optional(),
    singleAccountId: z.string().nullable().optional(),
    proxyAccountResponseCreateLimit: z.number().int().min(0).optional(),
    proxyAccountStreamLimit: z.number().int().min(0).optional(),
    proxyAccountStreamRecoveryReserve: z.number().int().min(0).optional(),
    proxyApiKeyFairShareCongestionThresholdPct: z.number().int().min(0).max(100).optional(),
    openaiCacheAffinityMaxAgeSeconds: z.number().int().positive().optional(),
    dashboardSessionTtlSeconds: z.number().int().min(3600).optional(),
    stickyReallocationBudgetThresholdPct: z.number().min(0).max(100).optional(),
    stickyReallocationPrimaryBudgetThresholdPct: z.number().min(0).max(100).optional(),
    stickyReallocationSecondaryBudgetThresholdPct: z.number().min(0).max(100).optional(),
    additionalQuotaRoutingPolicies: z
      .record(z.string(), AdditionalQuotaRoutingPolicySchema)
      .optional(),
    warmupModel: z.string().trim().min(1).optional(),
    importWithoutOverwrite: z.boolean().optional(),
    totpRequiredOnLogin: z.boolean().optional(),
    apiKeyAuthEnabled: z.boolean().optional(),
    hideUpstreamQuotaFromApiKeys: z.boolean().optional(),
    limitWarmupEnabled: z.boolean().optional(),
    limitWarmupWindows: LimitWarmupWindowsSchema.optional(),
    limitWarmupModel: LimitWarmupModelSchema.optional(),
    limitWarmupPrompt: LimitWarmupPromptSchema.optional(),
    limitWarmupCooldownSeconds: z.number().int().min(60).optional(),
    limitWarmupExhaustedThresholdPercent: z.number().positive().max(100).optional(),
    limitWarmupIdleThresholdPercent: z.number().positive().max(100).optional(),
    limitWarmupMinAvailablePercent: z.number().positive().max(100).optional(),
    weeklyPaceWorkingDays: WeeklyPaceWorkingDaysValueSchema.optional(),
    weeklyPaceSmoothingMinutes: WeeklyPaceSmoothingMinutesSchema.optional(),
    guestAccessEnabled: z.boolean().optional(),
    limitWarmupStaggeredIdleEnabled: z.boolean().optional(),
    // Tri-state overrides: absent = unchanged, null = clear (inherit env
    // alias), value = store the override.
    requestLogRetentionOverrideDays: z.number().int().min(0).max(3650).nullable().optional(),
    usageHistoryRetentionOverrideDays: z.number().int().min(0).max(3650).nullable().optional(),
  })
  .superRefine((settings, ctx) => {
    if (
      settings.proxyAccountStreamLimit !== undefined &&
      settings.proxyAccountStreamLimit > 0 &&
      settings.proxyAccountStreamRecoveryReserve !== undefined &&
      settings.proxyAccountStreamRecoveryReserve > settings.proxyAccountStreamLimit
    ) {
      ctx.addIssue({
        code: "custom",
        path: ["proxyAccountStreamRecoveryReserve"],
        message: "proxyAccountStreamRecoveryReserve must not exceed proxyAccountStreamLimit",
      });
    }
    if (
      settings.requestLogRetentionOverrideDays !== undefined &&
      settings.requestLogRetentionOverrideDays !== null &&
      settings.requestLogRetentionOverrideDays !== 0 &&
      settings.requestLogRetentionOverrideDays < 30
    ) {
      ctx.addIssue({
        code: "custom",
        path: ["requestLogRetentionOverrideDays"],
        message: "request_log_retention_override_days must be 0 (disabled) or >= 30",
      });
    }
    if (
      settings.usageHistoryRetentionOverrideDays !== undefined &&
      settings.usageHistoryRetentionOverrideDays !== null &&
      settings.usageHistoryRetentionOverrideDays !== 0 &&
      settings.usageHistoryRetentionOverrideDays < 45
    ) {
      ctx.addIssue({
        code: "custom",
        path: ["usageHistoryRetentionOverrideDays"],
        message: "usage_history_retention_override_days must be 0 (disabled) or >= 45",
      });
    }
  });

type ParsedDashboardSettings = z.infer<typeof DashboardSettingsSchema>;
type StickyThresholdPresenceFlags = Pick<
  ParsedDashboardSettings,
  | "__stickyReallocationBudgetThresholdPctProvided"
  | "__stickyReallocationPrimaryBudgetThresholdPctProvided"
  | "__stickyReallocationSecondaryBudgetThresholdPctProvided"
>;
type StickyThresholdValues = Pick<
  ParsedDashboardSettings,
  | "stickyReallocationBudgetThresholdPct"
  | "stickyReallocationPrimaryBudgetThresholdPct"
  | "stickyReallocationSecondaryBudgetThresholdPct"
>;

export type DashboardSettings = Omit<
  ParsedDashboardSettings,
  keyof StickyThresholdPresenceFlags | keyof StickyThresholdValues
> &
  Partial<StickyThresholdPresenceFlags> &
  Partial<StickyThresholdValues>;
export type SettingsUpdateRequest = z.infer<typeof SettingsUpdateRequestSchema>;
export type AdditionalQuotaRoutingPolicy = z.infer<typeof AdditionalQuotaRoutingPolicySchema>;

export const UpstreamProxyEndpointSchema = z.object({
  id: z.string(),
  name: z.string(),
  scheme: z.enum(["http", "https", "socks5", "socks5h"]),
  host: z.string(),
  port: z.number().int(),
  username: z.string().nullable().optional(),
  isActive: z.boolean(),
});

export const UpstreamProxyEndpointCreateRequestSchema = z.object({
  name: z.string().trim().min(1).max(128),
  scheme: z.enum(["http", "https", "socks5", "socks5h"]),
  host: z.string().trim().min(1).max(255),
  port: z.number().int().min(1).max(65535),
  username: z.string().trim().max(255).nullable().optional(),
  password: z.string().max(1024).nullable().optional(),
  isActive: z.boolean().optional().default(true),
});

export const UpstreamProxyEndpointUpdateRequestSchema = UpstreamProxyEndpointCreateRequestSchema;

export const UpstreamProxyEndpointTestResponseSchema = z.object({
  endpointId: z.string(),
  ok: z.boolean(),
  statusCode: z.number().int().nullable().optional(),
  elapsedMs: z.number().int().nullable().optional(),
  error: z.string().nullable().optional(),
});

export const UpstreamProxyPoolSchema = z.object({
  id: z.string(),
  name: z.string(),
  isActive: z.boolean(),
  endpointIds: z.array(z.string()),
});

export const UpstreamProxyPoolCreateRequestSchema = z.object({
  name: z.string().trim().min(1).max(128),
  endpointIds: z.array(z.string()).default([]),
  isActive: z.boolean().optional().default(true),
});

export const UpstreamProxyPoolUpdateRequestSchema = UpstreamProxyPoolCreateRequestSchema;

export const UpstreamProxyPoolMemberRequestSchema = z.object({
  endpointId: z.string().min(1),
  sortOrder: z.number().int().optional().default(0),
  weight: z.number().int().min(1).optional().default(1),
  isActive: z.boolean().optional().default(true),
});

export const AccountProxyBindingSchema = z.object({
  accountId: z.string(),
  poolId: z.string(),
  isActive: z.boolean(),
});

export const AccountProxyBindingRequestSchema = z.object({
  poolId: z.string().min(1),
  isActive: z.boolean().optional().default(true),
});

export const UpstreamProxyAdminSchema = z.object({
  routingEnabled: z.boolean(),
  defaultPoolId: z.string().nullable(),
  endpoints: z.array(UpstreamProxyEndpointSchema),
  pools: z.array(UpstreamProxyPoolSchema),
  bindings: z.array(AccountProxyBindingSchema),
});

export type UpstreamProxyEndpoint = z.infer<typeof UpstreamProxyEndpointSchema>;
export type UpstreamProxyEndpointCreateRequest = z.infer<typeof UpstreamProxyEndpointCreateRequestSchema>;
export type UpstreamProxyEndpointUpdateRequest = z.infer<typeof UpstreamProxyEndpointUpdateRequestSchema>;
export type UpstreamProxyEndpointTestResponse = z.infer<typeof UpstreamProxyEndpointTestResponseSchema>;
export type UpstreamProxyPool = z.infer<typeof UpstreamProxyPoolSchema>;
export type UpstreamProxyPoolCreateRequest = z.infer<typeof UpstreamProxyPoolCreateRequestSchema>;
export type UpstreamProxyPoolUpdateRequest = z.infer<typeof UpstreamProxyPoolUpdateRequestSchema>;
export type UpstreamProxyPoolMemberRequest = z.infer<typeof UpstreamProxyPoolMemberRequestSchema>;
export type AccountProxyBinding = z.infer<typeof AccountProxyBindingSchema>;
export type AccountProxyBindingRequest = z.infer<typeof AccountProxyBindingRequestSchema>;
export type UpstreamProxyAdmin = z.infer<typeof UpstreamProxyAdminSchema>;
